enhancements/cluster-capabilities-api.md

```diff
@@ -0,0 +1,337 @@
+creation-date: 2020-12-02
+last-updated: 2020-12-02
+status: implementable
+see-also:
+  - "/enhancements/single-node-production-edge-cluster-profile.md"
+  - "/enhancements/update/cluster-profiles.md"
+---
+
+# Cluster Capabilities API
+
+## Release Signoff Checklist
+
+- [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+As we add more deployment topologies for OpenShift, we need a way to
+communicate expectations to the operators managing the cluster. This
+enhancement describes an API for expressing the expected
+"capabilities" of a cluster, including aspects such as how hard
+operators should work to provide high-availability and whether
+operators should configure their operands in a way to minimize
+resource consumption for especially constrained environments.
+
+## Motivation
+
+We want to avoid having cluster profiles use different manifests to
+configure operators managed by the cluster-version-operator, and
+instead rely on in-cluster APIs. This approach will make the same
+information available to operators managed by either the
+cluster-version-operator or operator-lifecycle-manager.
+
+We do not want to expose the cluster profile name directly to most of
+the operators, because we do not want to have to update operators each
+time we add a new profile.
+
+We want the settings to be global to the cluster, rather than using
+operator-specific APIs, because that simplifies the integration with
+the installer and allows them to be applied consistently in all
+operators that need them.
+
+### Goals
+
+- Expose capability expectations to operators running in a cluster so
+  they can configure their operands appropriately.
+
+### Non-Goals
+
+- Expose cluster profile names to operators.
+- Provide a mutable API for changing the cluster capabilities (at
+  least for now).
+- Predict other capability settings that we might need to expose.
+
+## Proposal
+
+A new [informational
+resource](https://docs.openshift.com/container-platform/4.6/installing/install_config/customizations.html#informational-resources_customizations),
+`capabilities.config.openshift.io`, is created to hold details about
+the configured features of a cluster. It will include one field to
+start, and may be expanded over time.
+
+The `highAvailabilityMode` field will use an enum to express the
+high-availability expectations. Possible values are `full` and
+`none`. The default is `full`, which represents the behavior operators
+have today in a "normal" cluster. The `none` mode will be used in
+single-node deployments (developer and production), and the operator
+should not configure the operand for highly-available
+operation.
+
+The new cluster-wide `capabilities` API will supersede existing
```

Comment:
If we only have one capability, and it can fit in infrastructure, I don't see a reason to have the capabilities object yet.

Reply:
agree.

Reply:
We have identified 2 other candidates for capabilities.

One is a flag to indicate whether operators should deploy their GUI components. That could be part of an API that the console-operator defines, but the monitoring operator would need to use it and the installer would need to set it, so putting it in the capabilities seemed to make sense.

The other is some expression of the "size" operators should assume for their operands, so they could be reconfigured for smaller footprint deployments. We have not yet agreed on how to express that, and didn't want to hold this work up while we figured it out, so it was removed from this PR.

If those also seem like "infrastructure" settings, then we can rewrite this PR. But if we're going to have capabilities at some point, I would just as soon do it from the start.